Cloneable/CloneableGenerator.cs

```diff
@@ -41,77 +76,131 @@ private void GenerateCloneMethods(GeneratorExecutionContext context)
             if (context.SyntaxReceiver is not SyntaxReceiver receiver)
                 return;
 
-            var classSymbols = GetClassSymbols(context, receiver);
+            var options = (context.Compilation as CSharpCompilation).SyntaxTrees[0].Options as CSharpParseOptions;
+            
+            var compilation = context.Compilation.AddSyntaxTrees(CSharpSyntaxTree.ParseText(SourceText.From(cloneableAttributeText, Encoding.UTF8), options)).
+                AddSyntaxTrees(CSharpSyntaxTree.ParseText(SourceText.From(clonePropertyAttributeText, Encoding.UTF8), options)).
+                AddSyntaxTrees(CSharpSyntaxTree.ParseText(SourceText.From(ignoreClonePropertyAttributeText, Encoding.UTF8), options));
+
+            cloneableAttribute = compilation.GetTypeByMetadataName("Cloneable.CloneableAttribute")!;
+            cloneAttribute = compilation.GetTypeByMetadataName("Cloneable.CloneAttribute")!;
+            ignoreCloneAttribute = compilation.GetTypeByMetadataName("Cloneable.IgnoreCloneAttribute")!;
+
+            var classSymbols = GetClassSymbols(compilation, cloneableAttribute, receiver);
             foreach (var classSymbol in classSymbols)
             {
-                context.AddSource($"{classSymbol.Name}_cloneable.cs", SourceText.From(CreateCloneableCode(classSymbol), Encoding.UTF8));
+                var attribute = classSymbol.GetAttributes().First(x => x.AttributeClass.Equals(cloneableAttribute));
+                var isExplicit = (bool?)attribute.NamedArguments.FirstOrDefault(e => e.Key.Equals("ExplicitDeclaration")).Value.Value ?? false;
+                context.AddSource($"{classSymbol.Name}_cloneable.cs", SourceText.From(CreateCloneableCode(classSymbol, isExplicit), Encoding.UTF8));
             }
         }
 
-        private string CreateCloneableCode(INamedTypeSymbol classSymbol)
+        private string CreateCloneableCode(INamedTypeSymbol classSymbol, bool isExplicit)
         {
             string namespaceName = classSymbol.ContainingNamespace.ToDisplayString();
-            string fieldAssignmentsCode = GenerateFieldAssignmentsCode(classSymbol);
+            var fieldAssignmentsCode = GenerateFieldAssignmentsCode(classSymbol, isExplicit, false);
+            var fieldAssignmentsCodeSafe = fieldAssignmentsCode.Select(x =>
+            {
+                if (x.EndsWith("?.Clone"))
```

Comment:
Instead of checking the end of string, we can change the return type of GenerateFieldAssignmentsCode and return not the code but a class containing a property showing whether it is a deep copy or a shallow one.

Reply:
Check new code please.